keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,836 @@
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+### Notes/Constraints/Caveats (Optional)
+
+In changing the CRI, we must ensure that container runtime downstreams are able
+to support the new configurations.
+
+We considered adding parameters for both per-workload `memory-swap` and
+`swappiness`. These are documented as part of the Open Containers [runtime
+specification] for Linux memory configuration. Since `memory-swap` is a
+per-workload parameter, and `swappiness` is optional and can be set globally,
+we are choosing to only expose `memory-swap` which will adjust swap available
+to workloads.
+
+Since we are not currently setting `memory-swap` in the CRI, the current
+default behaviour when `--fail-swap-on=false` is set is to allocate the same
+amount of swap for a workload as memory requested. We will update the default
+to not permit the use of swap by setting `memory-swap` equal to `limit`.
+
+[runtime specification]: https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+
+### Risks and Mitigations
+
+Having swap available on a system reduces predictability. Swap's performance is
+worse than regular memory, sometimes by many orders of magnitude, which can
+cause unexpected performance regressions. Furthermore, swap changes a system's
+behaviour under memory pressure, and applications cannot directly control what
+portions of their memory usage are swapped out. Since enabling swap permits
+greater memory usage for workloads in Kubernetes that cannot be predictably
+accounted for, it also increases the risk of noisy neighbours and unexpected
+packing configurations, as the scheduler cannot account for swap memory usage.
+
+This risk is mitigated by preventing any workloads from using swap by default,
+even if swap is enabled and available on a system. This will allow a cluster
+administrator to test swap utilization just at the system level without
+introducing unpredictability to workload resource utilization.
+
+Additionally, we will mitigate this risk by determining a set of metrics to
+quantify system stability and then gathering test and production data to
+determine if system stability changes when swap is available to the system
+and/or workloads in a number of different scenarios.
+
+Since swap provisioning is out of scope of this proposal, this enhancement
+poses low risk to Kubernetes clusters that will not enable swap.
+
+## Design Details
+
+We summarize the implementation plan as following:
+
+1. Add a feature gate `NodeSwapEnabled` to enable swap support.
+1. Leave the default value of kubelet flag `--fail-on-swap` to `true`, to avoid
+   changing default behaviour.
+1. Introduce a new kubelet config parameter, `MemorySwapLimit`.
+1. Introduce a new CRI parameter, `memory_swap_limit_in_bytes`.
+1. Integrate new kubelet config and pass values to CRI for container creation.
```

Comment:
is any swap setting applied on each pod cgroup?  is the pod cgroup left unbounded?  if a container had used swap indirectly via a memory backed volume, and restarts, the charge will go to the pod cgroup, i am actually not sure what will happen here, but its worth clarifying if the pod cgroup is swap unbounded or not.  this would similarly impact ephemeral containers as well that have no defined resource requirements, are they able to consume swap?

Reply:
> is any swap setting applied on each pod cgroup? is the pod cgroup left unbounded? if a container had used swap indirectly via a memory backed volume, and restarts, the charge will go to the pod cgroup, i am actually not sure what will happen here, but its worth clarifying if the pod cgroup is swap unbounded or not.

As written, there are no cgroup changes in this proposal. I'll add that to alternatives, and we can consider adding that support in beta.

> this would similarly impact ephemeral containers as well that have no defined resource requirements, are they able to consume swap?

I don't actually know how this would behave right now. Since we're not setting cgroup-level restrictions, and ephemeral containers don't have a memory limit, I imagine the container runtimes permit the use of swap in this case, but I haven't tested it.

I'm thinking through how various different QoS classes will or should behave under swap. I think we can experiment for now but we'll need to concretely define how we want them to work for each swap configuration before graduating to beta. In particular, I'm not sure there is any way to prevent best effort workloads from using swap if it's available on a system.

Reply:
This needs a little more clarification.

The container runtime will write the swap settings to the container level cgroup.  This will include ephemeral containers.  We need to account for the proper setting that is applied at the pod cgroup boundary.  Please list addressing this as a blocker to beta for the feature